src/util.rs

```diff
@@ -97,6 +124,125 @@ impl Default for Digest {
+    /// pattern we're looking for is `__TIMESTAMP__` to avoid copying the entire
+    /// file to memory and re-searching the entire buffer for each pattern.
+    /// We can check inside each chunk for each pattern, and we use an overlap
+    /// buffer to keep the last `b"__TIMESTAMP__".len()` bytes around from the
+    /// last chunk, to also catch any pattern overlapping two chunks.
+    ///
+    /// In the above case, the overflow buffer would look like:
+    ///
+    /// ```text
+    ///    Chunk 1
+    ///    - aaaa0000
+    ///    Chunk 2
+    ///    - aaaabbbb
+    ///    - cccc0000
+    ///    Chunk 3
+    ///    - ccccdddd
+    ///    - eeee0000
+    ///    Chunk 4
+    ///    - eeeeffff
+    ///    [...]
+    /// ```
+    ///
+    /// We have to be careful to zero out the buffer right after each overlap check,
+    /// otherwise we risk the (unlikely) case of a pattern being spread between the
+    /// start of a chunk and its end.
+    /// There is also the edge case of the chunk (or even the entire file) being
+    /// smaller than a pattern to look for.
+    pub fn find_time_macros(&mut self, visit: &[u8]) {
+        if self.counter == 0 {
+            if visit.len() <= MAX_HAYSTACK_LEN {
+                // the file is smaller than the largest haystack
+                self.find_macros(visit);
+                self.counter += 1;
+                return;
+            }
+            // Copy the right side of the visit to the left of the buffer
+            let right_half = visit.len() - MAX_HAYSTACK_LEN;
+            self.overlap_buffer[..MAX_HAYSTACK_LEN].copy_from_slice(&visit[right_half..]);
+        } else {
+            if visit.len() < MAX_HAYSTACK_LEN {
+                // zero the right side of the buffer
+                self.overlap_buffer[MAX_HAYSTACK_LEN..].copy_from_slice(&[0; MAX_HAYSTACK_LEN]);
+                // Copy the visit to the right of the buffer, starting from the middle
+                self.overlap_buffer[MAX_HAYSTACK_LEN..MAX_HAYSTACK_LEN + visit.len()]
+                    .copy_from_slice(visit);
+            } else {
+                // Copy the left side of the visit to the right of the buffer
+                let left_half = MAX_HAYSTACK_LEN;
+                self.overlap_buffer[left_half..].copy_from_slice(&visit[..left_half]);
+                self.find_macros(&self.overlap_buffer.clone());
+                // zero the buffer
+                self.overlap_buffer = Default::default();
+                // Copy the right side of the visit to the left of the buffer
+                let right_half = visit.len() - MAX_HAYSTACK_LEN;
+                self.overlap_buffer[..MAX_HAYSTACK_LEN].copy_from_slice(&visit[right_half..]);
+            }
+            self.find_macros(&self.overlap_buffer.clone());
+        }
+
+        self.find_macros(visit);
+        self.counter += 1;
+    }
+
+    fn find_macros(&mut self, buffer: &[u8]) {
+        // TODO
```

Comment:
https://github.com/deepinthebuild/dedup/tree/master/fastchr might be an option

Reply:
`memchr` already uses SIMD when available and is probably the fastest generic approach. When I was referring to `ccache`'s use of SIMD, I was more thinking about their trick that all `__TIME__`, `__TIMESTAMP__` and `__DATE__` have the `__XXXE` mask, etc. It's more manual.